openedx/core/djangoapps/config_model_utils/models.py

```diff
@@ -0,0 +1,125 @@
+"""
+Useful ConfigurationModel subclasses
+
+StackedConfigurationModel: A ConfigurationModel that can be overridden at site, org and course levels
+"""
+
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from collections import namedtuple
+
+from django.conf import settings
+from django.db import models
+from django.contrib.sites.models import Site
+from django.utils.translation import ugettext_lazy as _
+
+from config_models.models import ConfigurationModel
+from openedx.core.djangoapps.site_configuration.models import SiteConfiguration
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+
+
+class StackedConfigurationModel(ConfigurationModel):
+    """
+    A ConfigurationModel that stacks Global, Site, Org, and Course level
+    configuration values.
+    """
+    class Meta(object):
+        abstract = True
+
+    KEY_FIELDS = ('site', 'org', 'course')
+
+    enabled = models.NullBooleanField(default=None, verbose_name=_("Enabled"))
+    site = models.ForeignKey(Site, on_delete=models.CASCADE, null=True)
+    org = models.CharField(max_length=255, db_index=True, null=True)
+    course = models.ForeignKey(
+        CourseOverview,
+        on_delete=models.DO_NOTHING,
+        null=True,
+    )
+
+    @classmethod
+    def attribute_tuple(cls):
+        "Returns a namedtuple with all attributes that can be overridden on this config model."
+        if hasattr(cls, '_attribute_tuple'):
+            return cls._attribute_tuple
+
+        cls._attribute_tuple = namedtuple(
+            '{}Values'.format(cls.__name__),
+            cls.STACKABLE_FIELDS
+        )
+        return cls._attribute_tuple
+
+    @classmethod
+    def current(cls, site=None, org=None, course=None):  # disable: pylint=arguments-differ
+        """
+        Return the current overridden configuration at the specified level.
+
+        Only one level may be specified at a time. Specifying multiple levels
+        will result in a ValueError.
+
+        Arguments:
+            site: The Site to check current values for
+            org: The org to check current values for
+            course: The course to check current values for
+
+        Returns:
+            An instance of :class:`cls.attribute_tuple()` with the overridden values
+            specified down to the level of the supplied argument (or global values if
+            no arguments are supplied).
+        """
+        if len([arg for arg in [site, org, course] if arg is not None]) > 1:
+            raise ValueError("Only one of site, org, and course can be specified")
```

Comment:
can you clarify how these config models will work?
suppose I want to enable content gating for 
site: edx.org
org: mit
org: harvard (Disabled)
course: python1
course: business1
will there be a separate config model for each of those things? 
 
and then to determine if the feature is enabled it will check the site/org/course for a particular request and ensure that site is true and org/course are true or None?
is this method just a utility method to check each individual config model but there will be future code to check them all together?

Reply:
Yeah, I'll add stuff to the docstring about that. In short: this method is the one that checks across all of the levels and aggregates the results. Maybe I shouldn't override `current`, since this kinda changes the semantics?

Reply:
> will there be a separate config model for each of those things?

There will be a separate model instance for each level of override that is defined.